pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,375 @@
+	c := &ExternalNodeController{
+		crdClient: crdClient,
+
+		externalNodeInformer:     externalNodeInformer,
+		externalNodeLister:       externalNodeInformer.Lister(),
+		externalNodeListerSynced: externalNodeInformer.Informer().HasSynced,
+
+		externalEntityInformer:     externalEntityInformer,
+		externalEntityLister:       externalEntityInformer.Lister(),
+		externalEntityListerSynced: externalEntityInformer.Informer().HasSynced,
+
+		syncedExternalNode: cache.NewStore(keyFunc),
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "externalnode"),
+	}
+	c.externalNodeInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.enqueueExternalNodeAdd,
+			UpdateFunc: c.enqueueExternalNodeUpdate,
+			DeleteFunc: c.enqueueExternalNodeDelete,
+		},
+		resyncPeriod)
+	return c
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeAdd(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode ADD event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeUpdate(oldObj interface{}, newObj interface{}) {
+	en := newObj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode UPDATE event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeDelete(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode DELETE event", "ExternalNode", klog.KObj(en))
+}
+
+// Run will create defaultWorkers workers (goroutines) which will process the ExternalEntity events from the work queue.
+func (c *ExternalNodeController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting", "controllerName", controllerName)
+	defer klog.InfoS("Shutting down", "controllerName", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalNodeListerSynced, c.externalEntityListerSynced) {
+		return
+	}
+	if err := c.reconcileExternalNodes(); err != nil {
+		klog.ErrorS(err, "Failed to reconcile ExternalNodes")
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// reconcileExternalNodes reconciles all the existing ExternalNodes and cleans up the stale ExternalEntities.
+func (c *ExternalNodeController) reconcileExternalNodes() error {
+	externalNodes, err := c.externalNodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	enUIDEENameMap := make(map[types.UID]string)
+	for _, en := range externalNodes {
+		if err = c.addExternalNode(en); err != nil {
+			return err
+		}
+		eeName := externalnode.GenExternalEntityName(en)
+		enUIDEENameMap[en.UID] = eeName
+	}
```

Comment:
Should we add existing nodes to syncedExternalNode?

Reply:
They will be added to syncedExternalNode inside function addExternalNode in line 151.

Reply:
Ok. I missed it.